packages/ai/core/tool/mcp.ts

```diff
@@ -0,0 +1,209 @@
+import { Client } from '@modelcontextprotocol/sdk/client/index.js';
+import {
+  StdioClientTransport,
+  StdioServerParameters,
+} from '@modelcontextprotocol/sdk/client/stdio.js';
+import {
+  SSEClientTransport,
+  SSEClientTransportOptions,
+} from '@modelcontextprotocol/sdk/client/sse.js';
+import {
+  ListToolsResult,
+  CallToolResultSchema,
+  CallToolResult,
+} from '@modelcontextprotocol/sdk/types';
+import { RequestOptions } from '@modelcontextprotocol/sdk/shared/protocol';
+
+import { jsonSchema, Schema } from '@ai-sdk/ui-utils';
+import { JSONSchema7 } from '@ai-sdk/provider';
+
+import { inferParameters, Tool } from './tool';
+import { z } from 'zod';
+
+type McpStdioServerConfig = StdioServerParameters & {
+  type: 'stdio';
+};
+interface McpSSEServerConfig {
+  type: 'sse';
+  url: string;
+  options?: SSEClientTransportOptions;
+}
+type McpServerConfig = McpStdioServerConfig | McpSSEServerConfig;
+
+interface CreateMcpToolsConfig {
+  servers: Record<string, McpServerConfig>;
+  toolCallOptions?: RequestOptions;
+}
+
+type McpCallToolInputSchema = z.ZodType<any, any, any> | Schema<any>;
+type McpToolSchemaMap = Record<string, { input: McpCallToolInputSchema }>;
+interface McpToolset<TToolSchemas extends McpToolSchemaMap = {}> {
+  tools: {
+    [K in keyof TToolSchemas]: Tool<
+      inferParameters<TToolSchemas[K]['input']>,
+      CallToolResult
+    >;
+  };
+  clients: Record<string, Client>;
+}
+
+// Based off: https://github.com/dnakov/claude-code/blob/da716d77b251868918bb5776377f85a18a47ffdf/src/services/mcpClient.ts#L224
+async function connectToServer(
+  name: string,
+  serverConfig: McpServerConfig,
+): Promise<Client> {
+  const transport =
+    serverConfig.type === 'sse'
+      ? new SSEClientTransport(new URL(serverConfig.url), serverConfig.options)
+      : new StdioClientTransport({
+          ...serverConfig,
+          env: {
+            ...process.env,
+            ...serverConfig.env,
+          } as Record<string, string>,
+        });
+
+  const client = new Client(
+    {
+      name: `${name}-client`,
+      version: '0.1.0',
+    },
+    {
+      capabilities: {},
```

Comment:
what can be set here?

Reply:
https://github.com/modelcontextprotocol/typescript-sdk/blob/170e656dcbb154b1e1723ba7dbb850d73010fbe1/src/types.ts#L196